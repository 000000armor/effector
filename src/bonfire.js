//@flow

import {type ID, createIDType} from './id'

const nextID = createIDType()
export class Bonfire<Field, State> {
  id: ID
  seq: number
  value: $ReadOnly<Field>
  defaultValue: $ReadOnly<Field>
  getter: (parent: State) => Field
  setter: (parent: State, value: Field) => State
}

export function init<Field, State>(
  defaultValue: Field,
  getter: (parent: State) => Field,
  setter: (parent: State, value: Field) => State
): Bonfire<Field, State> {
  const result = new Bonfire
  result.id = nextID()
  result.seq = 0
  result.value = defaultValue
  result.defaultValue = defaultValue
  result.getter = getter
  result.setter = setter
  return result
}

export function update<Field, State>(
  parent: State,
  updater: (
    // ( (parent: State, value: Field) => void ) |
    // ( (parent: State, value: Field) => Field )
    (parent: State, value: Field) => ( $NonMaybeType<Field> | void)
  ),
  bonfire: Bonfire<Field, State>
): {
  bonfire: Bonfire<Field, State>,
  parent: State,
} {
  const value = bonfire.getter(parent)
  const nextValueRaw = updater(parent, value)
  let nextValue: $NonMaybeType<Field>
  if (nextValueRaw === undefined)
    nextValue = bonfire.defaultValue
  else
    nextValue = nextValueRaw
  if (nextValue === value) return {
    bonfire,
    parent,
  }
  const nextState = bonfire.setter(parent, nextValue)
  const result = new Bonfire
  result.id = bonfire.id
  result.seq = bonfire.seq + 1
  result.value = nextValue
  return {
    bonfire: result,
    parent: nextState,
  }
}
