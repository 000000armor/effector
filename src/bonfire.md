# Bonfire

Bonfire is an instance of abstract reducer for the concrete place in state, relay. It wrap classic reducer and provide additional metainformation


## API

```typescript
type Bonfire<Field, State> = {
  id: ID,
  seq: number,
  value: Field,
  defaultValue: Field,
  getter: (parent: State) => Field,
  setter: (parent: State, value: Field) => State,
}
```
