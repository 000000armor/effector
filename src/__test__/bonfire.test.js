//@flow

import {init, update} from '../bonfire'

test('init', () => {
  expect(init(
    {field: 0},
    ({target}) => target,
    (parent, target) => ({
      ...parent,
      target,
    })
  )).toBeDefined()
})

describe('update', () => {
  const state = {
    target: {field: 0},
    another: 'foo',
  }

  const bonfire = init(
    {field: 0},
    ({target}) => target,
    (parent, target) => ({
      ...parent,
      target,
    })
  )
  test('should do immutable update', () => {
    expect(update(
      state,
      (parent, {field}) => ({field: field+1}),
      bonfire
    )).toMatchSnapshot()
  })
  test('should handle the same and undefined return value', () => {
    expect(update(
      state,
      () => {},
      bonfire
    )).toMatchSnapshot()

    expect(update(
      state,
      (parent, value) => value,
      bonfire
    ).parent).toBe(state)
  })
})
