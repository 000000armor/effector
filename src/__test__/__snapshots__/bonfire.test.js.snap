// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update should do immutable update 1`] = `
Object {
  "bonfire": Bonfire {
    "id": 1,
    "seq": 1,
    "value": Object {
      "field": 1,
    },
  },
  "parent": Object {
    "another": "foo",
    "target": Object {
      "field": 1,
    },
  },
}
`;

exports[`update should handle the same and undefined return value 1`] = `
Object {
  "bonfire": Bonfire {
    "id": 1,
    "seq": 1,
    "value": Object {
      "field": 0,
    },
  },
  "parent": Object {
    "another": "foo",
    "target": Object {
      "field": 0,
    },
  },
}
`;
